### PR TITLE
Migrate more micro-optimisations from Waterfall-Old

### PR DESCRIPTION
**Pull request by LuminousLoreBloom**
- Don't attempt to format arguments when there are none provided (-Mystiflow)
- Don't create a data input stream for every plugin message we get from servers  (-minecrafter)
- Optimize UUID dash replacements (-minecrafter)

### CODE REVIEW

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 1]
```diff
@@ -1,11 +1,28 @@
...
+diff --git a/api/src/main/java/net/md_5/bungee/api/ServerPing.java b/api/src/main/java/net/md_5/bungee/api/ServerPing.java
+index 27b5184..b854b9f 100644
+--- a/api/src/main/java/net/md_5/bungee/api/ServerPing.java
++++ b/api/src/main/java/net/md_5/bungee/api/ServerPing.java
+@@ -75,7 +75,8 @@ public class ServerPing
+ 
+         public String getId()
+         {
+-            return uniqueId.toString().replaceAll( "-", "" );
++            String id = uniqueId.toString();
```
Comment (jamierocks) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: final

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 2]
```diff
@@ -40,6 +72,40 @@ index e10ea44..5b02c5f 100644
...
+diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+index d87b03c..34e1737 100644
+--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
++++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+@@ -603,7 +603,8 @@ public class InitialHandler extends PacketHandler implements PendingConnection
+     @Override
+     public String getUUID()
+     {
+-        return uniqueId.toString().replaceAll( "-", "" );
++        String id = uniqueId.toString();
```
Comment (jamierocks) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: final

**Review by jamierocks — CHANGES_REQUESTED** [review]
👍

**Comment by LuminousLoreBloom** [comment]
Updated with requested changes

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 3]
```diff
@@ -27,8 +44,33 @@ index 71a5a15..520ee31 100644
...
+index a0f2c55..e270b66 100644
+--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
++++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+@@ -93,7 +93,8 @@ import org.fusesource.jansi.AnsiConsole;
+ /**
+  * Main BungeeCord proxy class.
+  */
+-public class BungeeCord extends ProxyServer
++public class
++BungeeCord extends ProxyServer
```
Comment (jamierocks) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: undo

**Review by jamierocks — CHANGES_REQUESTED** [review]
Just the one ;)

File: BungeeCord-Patches/0017-Micro-optimizations.patch (hunk 3 above)
Comment (LuminousLoreBloom) [reply BungeeCord-Patches/0017-Micro-optimizations.patch]: Whoops. Fixed. Also made further optimisations to ServerPing class.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 4]
```diff
@@ -1,13 +1,98 @@
...
+ import lombok.Data;
+-import lombok.Getter;
+ import lombok.NoArgsConstructor;
+ import lombok.ToString;
+ import net.md_5.bungee.Util;
+@@ -46,36 +45,54 @@ public class ServerPing
+     }
+ 
+     @Data
+-    @AllArgsConstructor
```
Comment (jamierocks) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Honestly I feel this is just going to be a pain to maintain over time.

**Review by jamierocks — CHANGES_REQUESTED** [review]

**Comment by Janmm14** [comment]
Isn't the final keyword for local variables just for source code clearness (if there are no anonymous things inside the method)?

**Comment by jamierocks** [comment]
> Isn't the final keyword for local variables just for source code clearness (if there are no anonymous things inside the method)?

Yep, just makes sure that no authors in the future do dumb stuff.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 5]
```diff
@@ -27,8 +112,55 @@ index 71a5a15..520ee31 100644
...
+diff --git a/proxy/src/main/java/net/md_5/bungee/PlayerInfoSerializer.java b/proxy/src/main/java/net/md_5/bungee/PlayerInfoSerializer.java
+index 491cf1a1..e02345b2 100644
+--- a/proxy/src/main/java/net/md_5/bungee/PlayerInfoSerializer.java
++++ b/proxy/src/main/java/net/md_5/bungee/PlayerInfoSerializer.java
+@@ -18,16 +18,8 @@ public class PlayerInfoSerializer implements JsonSerializer<ServerPing.PlayerInf
+     public ServerPing.PlayerInfo deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
+     {
+         JsonObject js = json.getAsJsonObject();
+-        ServerPing.PlayerInfo info = new ServerPing.PlayerInfo( js.get( "name" ).getAsString(), (UUID) null );
+-        String id = js.get( "id" ).getAsString();
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: What purpose does this change serve?

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 6]
```diff
@@ -40,6 +172,40 @@ index e10ea44..5b02c5f 100644
...
+index d87b03c0..1d341e2c 100644
+--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
++++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+@@ -603,7 +603,8 @@ public class InitialHandler extends PacketHandler implements PendingConnection
+     @Override
+     public String getUUID()
+     {
+-        return uniqueId.toString().replaceAll( "-", "" );
++        final String id = uniqueId.toString();
++        return new StringBuilder( 32 ).append( id, 0, 8 ).append( id, 9, 13 ).append( id, 14, 18 ).append( id, 19, 23 ).append( id, 24, 36 ).toString();
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Please refactor this out into a separate utility class.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 7]
```diff
@@ -1,13 +1,98 @@
...
+-            try
++            if ( undash )
+             {
+-                uniqueId = Util.getUUID( id );
+-            } catch ( Exception e )
+-            {
+-                // Fallback on a valid uuid otherwise Minecraft complains
+-                uniqueId = md5UUID;
++                undash = false;
++                uniqueId = new StringBuilder( 32 ).append( uniqueId, 0, 8 ).append( uniqueId, 9, 13 ).append( uniqueId, 14, 18 ).append( uniqueId, 19, 23 ).append( uniqueId, 24, 36 ).toString();
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Refractor this out into a separate utility method.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 8]
```diff
@@ -1,13 +1,98 @@
...
+ 
+         public PlayerInfo(String name, String id)
+         {
+-            setName( name );
+-            setId( id );
++            this.name = name;
++            this.uniqueId = id;
++            if ( id.indexOf( '-' ) != -1 )
++            {
++                this.undash = true;
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Instead of lazily un-dashing the uuid when getId is finally called, why not do it directly in the constructor?
Also, keep the logic for setting the uuid in the `setId` method, in case we ever need that.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 9]
```diff
@@ -1,13 +1,98 @@
...
++            }
+         }
+ 
+-        public void setId(String id)
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Why did you remove this method?

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 10]
```diff
@@ -1,13 +1,98 @@
...
+-    @AllArgsConstructor
+     public static class PlayerInfo
+     {
+-
+-        private String name;
+-        private UUID uniqueId;
++        
++        private final String name;
++        private String uniqueId;
++        private String dashedUuid;
```
Comment (Techcable) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Please directly store a `UUID` object instead of keeping a string object.

**Review by Techcable — CHANGES_REQUESTED** [review]
Although this seems like somewhat of a micro-optimization, I'd be willing to accept it if you can clean it up a bit :wink:

**Comment by Techcable** [comment]
@Janmm14 It also frees the JIT to do important optimizations, by assuming the variable remains unchanged after object initialization.
Immutable objects not only improve code quality, the  JIT can actually optimize them better than mutable data objects (especially in multi-threaded contexts).

File: BungeeCord-Patches/0017-Micro-optimizations.patch (hunk 5 above)
Comment (phase) [reply BungeeCord-Patches/0017-Micro-optimizations.patch]: It moved the dash checking to PlayerInfo.

File: BungeeCord-Patches/0017-Micro-optimizations.patch (hunk 10 above)
Comment (phase) [reply BungeeCord-Patches/0017-Micro-optimizations.patch]: I think the entire point of this micro-optimization is to remove the UUID objects.

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 11]
```diff
@@ -1,13 +1,98 @@
...
++        private boolean undash;
+ 
+         private static final UUID md5UUID = Util.getUUID( "af74a02d19cb445bb07f6866a861f783" );
++        private static final String md5ID = md5UUID.toString();
```
Comment (phase) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Don't make a UUID object you don't need.

```java
private static final String md5UUID = "af74a02d19cb445bb07f6866a861f783";
```

Comment (Janmm14) [reply BungeeCord-Patches/0017-Micro-optimizations.patch]: with dashes please

**Comment by LuminousLoreBloom** [comment]
I agree with jamierocks, I'm just going to discard the majority of the UUID PlayerInfo code as I also feel there's too much diff for negligible gain.

I should have the PR updated with other suggested changes in a bit.

File: BungeeCord-Patches/0011-Optimize-uuid-conversions.patch [hunk 12]
```diff
@@ -240,6 +244,51 @@ index 86a0055..6c9c6d6 100644
...
+ import lombok.AllArgsConstructor;
+ import lombok.Data;
+ import lombok.Getter;
+@@ -75,7 +77,8 @@ public class ServerPing
+ 
+         public String getId()
+         {
+-            return uniqueId.toString().replaceAll( "-", "" );
++
++            return UUIDUtils.undash( uniqueId.toString() );
```
Comment (jamierocks) [inline BungeeCord-Patches/0011-Optimize-uuid-conversions.patch]: Could you use the fully-qualified class name here, to remove the need for the import?

File: BungeeCord-Patches/0011-Optimize-uuid-conversions.patch [hunk 13]
```diff
@@ -240,6 +244,51 @@ index 86a0055..6c9c6d6 100644
...
++import io.github.waterfallmc.waterfall.utils.UUIDUtils;
+ import lombok.Getter;
+ import lombok.RequiredArgsConstructor;
+ import net.md_5.bungee.BungeeCord;
+@@ -603,7 +605,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
+     @Override
+     public String getUUID()
+     {
+-        return uniqueId.toString().replaceAll( "-", "" );
++        return UUIDUtils.undash( uniqueId.toString() );
```
Comment (jamierocks) [inline BungeeCord-Patches/0011-Optimize-uuid-conversions.patch]: Could you use the fully-qualified class name here, to remove the need for the import?

File: BungeeCord-Patches/0017-Micro-optimizations.patch [hunk 14]
```diff
@@ -1,13 +1,17 @@
...
+From ff9ecc674b6bf263f8b8d4219b5692b9e06bb868 Mon Sep 17 00:00:00 2001
 From: Tux <write@imaginarycode.com>
 Date: Thu, 19 May 2016 18:05:33 -0600
 Subject: [PATCH] Micro-optimizations
 
 - PluginManager.dispatchCommand() avoids regex while splitting commands. Java 7 introduced an optimized String.split() that should be used instead (affects command dispatch).
 - Avoid regex in getLocale() by changing from replaceAll(String, String) to replaceAll(char, char)
+- Don't attempt to format arguments when there are none provided
+- Don't create a data input stream for every plugin message we get from
+servers
```
Comment (jamierocks) [inline BungeeCord-Patches/0017-Micro-optimizations.patch]: Why two lines?

**Review by jamierocks — CHANGES_REQUESTED** [review]
:+1:

File: BungeeCord-Patches/0011-Optimize-uuid-conversions.patch (hunk 12 above)
Comment (jamierocks) [reply BungeeCord-Patches/0011-Optimize-uuid-conversions.patch]: And why the line above it?

**Comment by LuminousLoreBloom** [comment]
Updated to use fully-qualified class imports and fixed mistakes pointed out.

**Review by jamierocks — APPROVED** [review]

**Comment by jamierocks** [comment]
Cheers ;)
